### PR TITLE
[DistillationTrainer refactor] Freeze ServerDistillationTrainer self-contained

### PR DESCRIPTION
**Pull request by qgallouedec**
*Item 21b (Group B) — stacked on PR 21. Part of #6449. ⚠️ sequencing hazard (H7).*

`ServerDistillationTrainer` depended on base internals that items 22/29/33 delete (`_jsd_divergence`, `_reduce_divergence_loss`, `_compute_prompt_length`) and on the old `input_ids`/`labels` layout (dropped at 29). This decouples it so the base deletions keep it green; the full re-point onto the stable base stays item 55.

- Own copies of `_jsd_divergence` (module fn), `_compute_prompt_length`, `_reduce_divergence_loss` (verbatim from the base) — no more base import of the soon-deleted helpers.
- `compute_loss` reconstructs `input_ids`/`attention_mask`/`labels`/`prompt_attention_mask` from GRPO's `prompt_ids`/`completion_ids`/`prompt_mask`/`completion_mask`, so it rides on the base's (evolving) generation via the new keys.
- Behaviour-preserving: the reconstruction equals the emitted old-layout tensors (`cat(prompt_ids, completion_ids) == input_ids`, etc.). Server test import switched to the server's `_jsd_divergence`.

Verified: 12 server tests passed on the core change; ruff clean. (Final re-run pending an unrelated env import hang in `trl.trainer.utils`.)

---

> [!NOTE]
> **Medium Risk**
> Touches training loss computation with verbatim-copied divergence math; intended behavior-preserving but errors in tensor reconstruction would affect gradients.
> 
> **Overview**
> **Decouples** `ServerDistillationTrainer` from base `DistillationTrainer` helpers that upcoming distillation refactors will remove, so server distillation keeps working while the base migrates to chunked loss (#6449).
> 
> The server module now **vendors** `_jsd_divergence` (module-level), `_compute_prompt_length`, and `_reduce_divergence_loss` instead of importing them from the base trainer. **`compute_loss`** rebuilds the legacy `input_ids` / `attention_mask` / `labels` / `prompt_attention_mask` tensors from GRPO-style `prompt_ids`, `completion_ids`, `prompt_mask`, and `completion_mask` before the existing teacher-server loss path runs.
> 
> Tests import `_jsd_divergence` from the server trainer module rather than the base distillation trainer.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a37731af76939e79605c35a3bbe8945eba784fab. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/server_distillation/server_distillation_trainer.py [hunk 1]
```diff
@@ -194,7 +241,66 @@ def _compute_sparse_top_1_divergence_loss(
...
+            return jsd.mean()
+        else:
+            return jsd
+
     def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
+        # Reconstruct the `input_ids` / `attention_mask` / `labels` layout this server path is built on from GRPO's
+        # `prompt_ids` / `completion_ids` / `completion_mask` keys, so it rides on the base's (evolving) generation. The
+        # teacher-server loss is migrated onto the stable base wholesale later (issue #6449, item 58).
+        input_ids = torch.cat([inputs["prompt_ids"], inputs["completion_ids"]], dim=1)
+        attention_mask = torch.cat([inputs["prompt_mask"], inputs["completion_mask"]], dim=1)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/server_distillation/server_distillation_trainer.py]: Good call. Aligned. The shared reconstruction now mirrors DistillationTrainer.compute_loss verbatim the server-only labels/prompt_attention_mask rebuild is separated below with its own comment.

**Review by cmpatino — APPROVED** [review]
